Support generation of Job/CronJob resources

Sgitario opened the pull request:
Fix https://github.com/quarkusio/quarkus/issues/27024

gsmet commented on docs/src/main/asciidoc/deploying-to-kubernetes.adoc:
```diff
@@ -175,6 +168,47 @@ quarkus.container-image.tag=1.0       #optional, defaults to the application ver
 
+=== Changing the generated deployment resource
+
+Beside generating a `Deployment` resource, you can also choose to get either a `StatefulSet`, or a `Job`, or a `CronJob` resource instead via `application.properties`:
```
```suggestion
Besides generating a `Deployment` resource, you can also choose to generate either a `StatefulSet`, or a `Job`, or a `CronJob` resource instead via `application.properties`:
```
`beside` and `besides` both exists but given it's either one or another, I think it should be `besides`.

gsmet commented on docs/src/main/asciidoc/deploying-to-kubernetes.adoc:
```diff
@@ -175,6 +168,47 @@ quarkus.container-image.tag=1.0       #optional, defaults to the application ver
+
+==== Generating Job resources
+
+If you want to generate a Job resource, you need to add the following property via the `application.properties`:
```
```suggestion
If you want to generate a Job resource, you need to add the following property to the `application.properties`:
```

gsmet commented on docs/src/main/asciidoc/deploying-to-kubernetes.adoc:
```diff
@@ -175,6 +168,47 @@ quarkus.container-image.tag=1.0       #optional, defaults to the application ver
+quarkus.kubernetes.deployment-kind=Job
+----
+
+IMPORTANT: If you are using the Picocli extension, by default the Job resource will be generated.
```
```suggestion
IMPORTANT: If you are using the Picocli extension, by default a Job resource will be generated.
```

gsmet commented on docs/src/main/asciidoc/deploying-to-kubernetes.adoc:
```diff
@@ -175,6 +168,47 @@ quarkus.container-image.tag=1.0       #optional, defaults to the application ver
+
+IMPORTANT: If you are using the Picocli extension, by default the Job resource will be generated.
+
+You can provide the arguments that will be used by the Kubernetes Job via the property `quarkus.kubernetes.arguments`. For example, adding the property `quarkus.kubernetes.arguments=A,B`.
```
```suggestion
You can provide the arguments that will be used by the Kubernetes Job via the property `quarkus.kubernetes.arguments`. For example, by adding the property `quarkus.kubernetes.arguments=A,B`.
```

gsmet commented on docs/src/main/asciidoc/deploying-to-kubernetes.adoc:
```diff
@@ -175,6 +168,47 @@ quarkus.container-image.tag=1.0       #optional, defaults to the application ver
+
+You can provide the arguments that will be used by the Kubernetes Job via the property `quarkus.kubernetes.arguments`. For example, adding the property `quarkus.kubernetes.arguments=A,B`.
+
+Finally, the Kubernetes job will be launched every time that is installed in Kubernetes. You can know more about how to run Kubernetes jobs in this https://kubernetes.io/docs/concepts/workloads/controllers/job/#running-an-example-job[link].
```
```suggestion
Finally, the Kubernetes job will be launched every time it is installed in Kubernetes. You can know more about how to run Kubernetes jobs in this https://kubernetes.io/docs/concepts/workloads/controllers/job/#running-an-example-job[link].
```

gsmet commented on docs/src/main/asciidoc/deploying-to-kubernetes.adoc:
```diff
@@ -175,6 +168,47 @@ quarkus.container-image.tag=1.0       #optional, defaults to the application ver
+
+Finally, the Kubernetes job will be launched every time that is installed in Kubernetes. You can know more about how to run Kubernetes jobs in this https://kubernetes.io/docs/concepts/workloads/controllers/job/#running-an-example-job[link].
+
+You can configure the rest of the Kubernetes Job configuration using the properties under `quarkus.kubernetes.job.xxx` (see https://quarkus.io/guides/deploying-to-kubernetes#quarkus-kubernetes-kubernetes-config_quarkus.kubernetes.job.parallelism-parallelism[link]).
```
```suggestion
You can configure the rest of the Kubernetes Job configuration using the properties under `quarkus.kubernetes.job.xxx` (see the https://quarkus.io/guides/deploying-to-kubernetes#quarkus-kubernetes-kubernetes-config_quarkus.kubernetes.job.parallelism-parallelism[Deploying to Kubernetes guide] for more information).
```

gsmet commented on docs/src/main/asciidoc/deploying-to-kubernetes.adoc:
```diff
@@ -175,6 +168,47 @@ quarkus.container-image.tag=1.0       #optional, defaults to the application ver
+
+==== Generating CronJob resources
+
+If you want to generate a CronJob resource, you need to add the following property via the `application.properties`:
```
```suggestion
If you want to generate a CronJob resource, you need to add the following properties to the `application.properties`:
```

gsmet commented on docs/src/main/asciidoc/deploying-to-kubernetes.adoc:
```diff
@@ -175,6 +168,47 @@ quarkus.container-image.tag=1.0       #optional, defaults to the application ver
+quarkus.kubernetes.cron-job.schedule=0 * * * *
+----
+
+IMPORTANT: CronJob resources require the https://en.wikipedia.org/wiki/Cron[Cron] expression to specify when to launch the job via the property `quarkus.kubernetes.cron-job.schedule`. If not provide, the build will fail.
```
```suggestion
IMPORTANT: CronJob resources require the https://en.wikipedia.org/wiki/Cron[Cron] expression to specify when to launch the job via the property `quarkus.kubernetes.cron-job.schedule`. If not provided, the build will fail.
```

gsmet commented on docs/src/main/asciidoc/deploying-to-kubernetes.adoc:
```diff
@@ -175,6 +168,47 @@ quarkus.container-image.tag=1.0       #optional, defaults to the application ver
+
+IMPORTANT: CronJob resources require the https://en.wikipedia.org/wiki/Cron[Cron] expression to specify when to launch the job via the property `quarkus.kubernetes.cron-job.schedule`. If not provide, the build will fail.
+
+You can configure the rest of the Kubernetes CronJob configuration using the properties under `quarkus.kubernetes.cron-job.xxx` (see https://quarkus.io/guides/deploying-to-kubernetes#quarkus-kubernetes-kubernetes-config_quarkus.kubernetes.cron-job.parallelism-parallelism[link]).
```
```suggestion
You can configure the rest of the Kubernetes CronJob configuration using the properties under `quarkus.kubernetes.cron-job.xxx` (see the https://quarkus.io/guides/deploying-to-kubernetes#quarkus-kubernetes-kubernetes-config_quarkus.kubernetes.cron-job.parallelism-parallelism[Deploying to Kubernetes guide] for more information).
```

gsmet commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -354,7 +354,16 @@ It's also possible to use the value from another field to add a new environment
-==== Using Deployment instead of DeploymentConfig
+==== Changing the generated deployment resource
+
+Beside generating a `DeploymentConfig` resource, you can also choose to get either a `Deployment`, `StatefulSet`, or a `Job`, or a `CronJob` resource instead via `application.properties`:
```
```suggestion
Besides generating a `DeploymentConfig` resource, you can also choose to generate either a `Deployment`, `StatefulSet`, or a `Job`, or a `CronJob` resource instead via `application.properties`:
```

gsmet commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -371,6 +380,38 @@ When the image is built, using OpenShift builds (s2i binary and docker strategy)
 
+===== Generating Job resources
+
+If you want to generate a Job resource, you need to add the following property via the `application.properties`:
```
```suggestion
If you want to generate a Job resource, you need to add the following property to the `application.properties`:
```

gsmet commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -371,6 +380,38 @@ When the image is built, using OpenShift builds (s2i binary and docker strategy)
+quarkus.openshift.deployment-kind=Job
+----
+
+IMPORTANT: If you are using the Picocli extension, by default the Job resource will be generated.
```
```suggestion
IMPORTANT: If you are using the Picocli extension, by default a Job resource will be generated.
```

gsmet commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -371,6 +380,38 @@ When the image is built, using OpenShift builds (s2i binary and docker strategy)
+
+IMPORTANT: If you are using the Picocli extension, by default the Job resource will be generated.
+
+You can provide the arguments that will be used by the Kubernetes Job via the property `quarkus.openshift.arguments`. For example, adding the property `quarkus.openshift.arguments=A,B`.
```
```suggestion
You can provide the arguments that will be used by the Kubernetes Job via the property `quarkus.openshift.arguments`. For example, by adding the property `quarkus.openshift.arguments=A,B`.
```

gsmet commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -371,6 +380,38 @@ When the image is built, using OpenShift builds (s2i binary and docker strategy)
+
+You can provide the arguments that will be used by the Kubernetes Job via the property `quarkus.openshift.arguments`. For example, adding the property `quarkus.openshift.arguments=A,B`.
+
+Finally, the Kubernetes job will be launched every time that is installed in OpenShift. You can know more about how to run Kubernetes jobs in this https://kubernetes.io/docs/concepts/workloads/controllers/job/#running-an-example-job[link].
```
```suggestion
Finally, the Kubernetes job will be launched every time it is installed in OpenShift. You can know more about how to run Kubernetes jobs in this https://kubernetes.io/docs/concepts/workloads/controllers/job/#running-an-example-job[document].
```

gsmet commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -371,6 +380,38 @@ When the image is built, using OpenShift builds (s2i binary and docker strategy)
+
+Finally, the Kubernetes job will be launched every time that is installed in OpenShift. You can know more about how to run Kubernetes jobs in this https://kubernetes.io/docs/concepts/workloads/controllers/job/#running-an-example-job[link].
+
+You can configure the rest of the Kubernetes Job configuration using the properties under `quarkus.openshift.job.xxx` (see https://quarkus.io/guides/deploying-to-openshift#quarkus-openshift-openshift-config_quarkus.openshift.job.parallelism[link]).
```
```suggestion
You can configure the rest of the Kubernetes Job configuration using the properties under `quarkus.openshift.job.xxx` (see the https://quarkus.io/guides/deploying-to-openshift#quarkus-openshift-openshift-config_quarkus.openshift.job.parallelism[Deploying to OpenShift guide] for more information).
```

gsmet commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -371,6 +380,38 @@ When the image is built, using OpenShift builds (s2i binary and docker strategy)
+
+===== Generating CronJob resources
+
+If you want to generate a CronJob resource, you need to add the following property via the `application.properties`:
```
```suggestion
If you want to generate a CronJob resource, you need to add the following properties to the `application.properties`:
```

gsmet commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -371,6 +380,38 @@ When the image is built, using OpenShift builds (s2i binary and docker strategy)
+quarkus.openshift.cron-job.schedule=0 * * * *
+----
+
+IMPORTANT: CronJob resources require the https://en.wikipedia.org/wiki/Cron[Cron] expression to specify when to launch the job via the property `quarkus.openshift.cron-job.schedule`. If not provide, the build will fail.
```
```suggestion
IMPORTANT: CronJob resources require the https://en.wikipedia.org/wiki/Cron[Cron] expression to specify when to launch the job via the property `quarkus.openshift.cron-job.schedule`. If not provided, the build will fail.
```

gsmet commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -371,6 +380,38 @@ When the image is built, using OpenShift builds (s2i binary and docker strategy)
+
+IMPORTANT: CronJob resources require the https://en.wikipedia.org/wiki/Cron[Cron] expression to specify when to launch the job via the property `quarkus.openshift.cron-job.schedule`. If not provide, the build will fail.
+
+You can configure the rest of the Kubernetes CronJob configuration using the properties under `quarkus.openshift.cron-job.xxx` (see https://quarkus.io/guides/deploying-to-openshift#quarkus-openshift-openshift-config_quarkus.openshift.cron-job.parallelism[link]).
```
```suggestion
You can configure the rest of the Kubernetes CronJob configuration using the properties under `quarkus.openshift.cron-job.xxx` (see the https://quarkus.io/guides/deploying-to-openshift#quarkus-openshift-openshift-config_quarkus.openshift.cron-job.parallelism[Deploying to OpenShift guide] for more information).
```

gsmet commented on docs/src/main/asciidoc/picocli.adoc:
```diff
@@ -291,6 +291,81 @@ annotationProcessor 'info.picocli:picocli-codegen'
 
+== Kubernetes support
+
+Once you have your command line application, you can also generate the resources necessary to install and use this application in Kubernetes by adding the `kubernetes` extension. To install the `kubernetes` extension, run the following command in your project base directory.
```
```suggestion
Once you have your command line application, you can also generate the resources necessary to install and use this application in Kubernetes by adding the `kubernetes` extension. To install the `kubernetes` extension, run the following command in your project base directory:
```

gsmet commented on docs/src/main/asciidoc/picocli.adoc:
```diff
@@ -291,6 +291,81 @@ annotationProcessor 'info.picocli:picocli-codegen'
+
+Once you have your command line application, you can also generate the resources necessary to install and use this application in Kubernetes by adding the `kubernetes` extension. To install the `kubernetes` extension, run the following command in your project base directory.
+
+[source,bash]
+----
+./mvnw quarkus:add-extension -Dextensions="kubernetes"
+----
+
+This will add the following to your `pom.xml`:
+
+[source,xml]
+----
+<dependency>
+    <groupId>io.quarkus</groupId>
+    <artifactId>quarkus-kubernetes</artifactId>
+</dependency>
+----
```
Ideally you should use the tab thing here to have Maven/CLI and Gradle/Maven alternatives. See other guides for examples.

Sgitario replied:
Awesome! I also changed the other one in the same guide for picocli.

gsmet commented on docs/src/main/asciidoc/picocli.adoc:
```diff
@@ -291,6 +291,81 @@ annotationProcessor 'info.picocli:picocli-codegen'
+</dependency>
+----
+
+And, next, build the application following the execution of:
```
```suggestion
And, next, build the application with:
```

gsmet commented on docs/src/main/asciidoc/picocli.adoc:
```diff
@@ -291,6 +291,81 @@ annotationProcessor 'info.picocli:picocli-codegen'
+
+include::{includes}/devtools/build.adoc[]
+
+The Kubernetes extension will detect that the presence of the Picocli extension and hence generate a https://kubernetes.io/docs/concepts/workloads/controllers/job/[Job] resource instead of a https://kubernetes.io/docs/concepts/workloads/controllers/deployment/[Deployment] resource in the `target/kubernetes/` directory.
```
```suggestion
The Kubernetes extension will detect the presence of the Picocli extension and hence generate a https://kubernetes.io/docs/concepts/workloads/controllers/job/[Job] resource instead of a https://kubernetes.io/docs/concepts/workloads/controllers/deployment/[Deployment] resource in the `target/kubernetes/` directory.
```

gsmet commented on docs/src/main/asciidoc/picocli.adoc:
```diff
@@ -291,6 +291,81 @@ annotationProcessor 'info.picocli:picocli-codegen'
+
+IMPORTANT: If you don't want to generate a Job resource, you can specify the resource you want to generate using the property `quarkus.kubernetes.deployment-kind`. For example, if you want to generate a Deployment resource, use `quarkus.kubernetes.deployment-kind=Deployment`.
+
+Moreover, you can provide the arguments that will be used by the Kubernetes job via the property `quarkus.kubernetes.arguments`. For example, adding the property `quarkus.kubernetes.arguments=A,B` and after building your project, the following Job resource will be generated:
```
```suggestion
Moreover, you can provide the arguments that will be used by the Kubernetes job via the property `quarkus.kubernetes.arguments`. For example, after adding the property `quarkus.kubernetes.arguments=A,B` and building your project, the following Job resource will be generated:
```

gsmet commented on docs/src/main/asciidoc/picocli.adoc:
```diff
@@ -291,6 +291,81 @@ annotationProcessor 'info.picocli:picocli-codegen'
+
+----
+
+Finally, the Kubernetes job will be launched every time that is installed in Kubernetes. You can know more about how to run Kubernetes jobs in this https://kubernetes.io/docs/concepts/workloads/controllers/job/#running-an-example-job[link].
```
```suggestion
Finally, the Kubernetes job will be launched every time it is installed in Kubernetes. You can know more about how to run Kubernetes jobs in this https://kubernetes.io/docs/concepts/workloads/controllers/job/#running-an-example-job[document].
```

gsmet reviewed (changes requested):
I haven't reviewed the code, only the documentation.

gsmet commented:
@Sgitario I think you missed some of my suggestions. GitHub displays some of them in the Conversation tab but it hides those in the middle and you have to click on `14 hidden conversations Load more`.

Sgitario commented:
> @Sgitario I think you missed some of my suggestions. GitHub displays some of them in the Conversation tab but it hides those in the middle and you have to click on `14 hidden conversations Load more`.

Ops, sorry! All the suggestions are now applied.

gsmet commented:
OK, I'll take it from there as I want to apply one more tab related change.

gsmet commented:
I squashed the commits and also used the tabs for pom.xml/build.gradle alternatives in the Picocli doc.

gsmet reviewed (approved):
I'm good on the doc side, I will let @iocanel check the code itself.

geoand commented:
@iocanel mind taking a look at this?

iocanel reviewed (changes requested):
we need to check that working with existing resources dont throw CME. A test maybe?

Sgitario committed:
Support generation of Job/CronJob resources

Fix https://github.com/quarkusio/quarkus/issues/27024

Sgitario commented:
@iocanel as far I know, the decorators are applied sequentially, so don't think we should worry about concurrent modifications. Moreover, we're doing the same in other decorators. However, to be extra safe, I've added this test https://github.com/quarkusio/quarkus/blob/389d91b8170a30248954788d41201f80fc8c8c31/integration-tests/kubernetes/quarkus-standard-way/src/test/java/io/quarkus/it/kubernetes/KubernetesWithExistingCronJobResourceTest.java. 

Btw, after rebasing the branch, I spotted an issue when applying the matching labels. I've already fixed it, but I need to fix it in Dekorate as well.

iocanel commented:
> @iocanel as far I know, the decorators are applied sequentially, so don't think we should worry about concurrent modifications. Moreover, we're doing the same in other decorators. However, to be extra safe, I've added this test

It doesn'matter how you apply the decorators. Modifying a list while iterating it will result in this exception. I think the same applies to stream. Eitherway, having a test is good enough for me.

iocanel reviewed (approved):
lgtm

Sgitario commented:
Tests failures are unrelated. I'm merging this.
